tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java

```diff
@@ -143,22 +396,134 @@ public void testListGroupCommand(String quorum) throws Exception {
+                return null;
+            }));
+            return out.get().contains("STATE") && !out.get().contains("TYPE") && out.get().contains(simpleGroup) && out.get().contains(PROTOCOL_GROUP);
+        }, "Expected to find " + simpleGroup + ", " + PROTOCOL_GROUP + " and state header, but found " + out.get());
+
+        String[] cgcArgs3 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--type"};
+        TestUtils.waitForCondition(() -> {
+            out.set(kafka.utils.TestUtils.grabConsoleOutput(() -> {
+                ConsumerGroupCommand.main(cgcArgs3);
+                return null;
+            }));
+            return out.get().contains("TYPE") && out.get().contains("Consumer") && !out.get().contains("STATE") &&
+                out.get().contains(PROTOCOL_GROUP) && out.get().contains(simpleGroup);
+        }, "Expected to find " + simpleGroup + ", " + PROTOCOL_GROUP + " and type header, but found " + out.get());
+
+        String[] cgcArgs4 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--type", "consumer"};
+        TestUtils.waitForCondition(() -> {
+            out.set(kafka.utils.TestUtils.grabConsoleOutput(() -> {
+                ConsumerGroupCommand.main(cgcArgs4);
+                return null;
+            }));
+            return out.get().contains("TYPE") && out.get().contains("Consumer") && !out.get().contains("STATE") && out.get().contains(PROTOCOL_GROUP);
+        }, "Expected to find " + PROTOCOL_GROUP + " with type header and Consumer type, but found " + out.get());
+
+        String[] cgcArgs5 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--type", "consumer", "--state", "Stable"};
+        TestUtils.waitForCondition(() -> {
+            out.set(kafka.utils.TestUtils.grabConsoleOutput(() -> {
+                ConsumerGroupCommand.main(cgcArgs5);
+                return null;
+            }));
+            return out.get().contains("TYPE") && out.get().contains("Consumer") && out.get().contains("STATE") &&
+                out.get().contains("Stable") && out.get().contains(PROTOCOL_GROUP);
+        }, "Expected to find " + PROTOCOL_GROUP + " with type and state header, but found " + out.get());
+    }
+
+    private static void assertGroupListing(
+        ConsumerGroupCommand.ConsumerGroupService service,
+        Set<GroupType> typeFilterSet,
+        Set<ConsumerGroupState> stateFilterSet,
+        Set<ConsumerGroupListing> expectedListing
+    ) throws Exception {
+        final AtomicReference<scala.collection.Set> foundListing = new AtomicReference<>();
+        TestUtils.waitForCondition(() -> {
+            foundListing.set(service.listConsumerGroupsWithFilters(set(typeFilterSet), set(stateFilterSet)).toSet());
+            return Objects.equals(set(expectedListing), foundListing.get());
+        }, "Expected to show groups " + expectedListing + ", but found " + foundListing.get() + ".");
```

Comment:
I just reaslized that there is a small bug here. It must be `() -> "Expected...` because `foundListing` is always null otherwise.

Reply:
Oh yeah thanks for pointing this out! I didn't realise